File: typescript/src/parsers/variableParser.ts
/**
 * Parser for Terraform variable blocks.
 * Extracts variable declarations with type constraints, defaults, and validations.
 */

import { HclBlock, TypeConstraint, VariableBlock, VariableValidation } from '../types/blocks';
import { parseBlockBody } from '../utils/parser/bodyParser';
import { literalBoolean, literalString } from '../utils/common/valueHelpers';

/**
 * Parser for Terraform variable definition blocks.
 *
 * @example
 * ```hcl
 * variable "instance_type" {
 *   type        = string
 *   default     = "t2.micro"
 *   description = "EC2 instance type"
 *   sensitive   = false
 *
 *   validation {
 *     condition     = can(regex("^t[23]\\.", var.instance_type))
 *     error_message = "Must be a t2 or t3 instance type."
 *   }
 * }
 * ```
 */
export class VariableParser {
    /**
     * Parses a variable block into a structured VariableBlock.
     * @param block - The raw HCL block to parse
     * @returns Parsed VariableBlock with all extracted fields
     */
    parse(block: HclBlock): VariableBlock {
        const name = block.labels[0] || 'unknown';
        const parsed = parseBlockBody(block.body);

        const description = literalString(parsed.attributes.description) ?? parsed.attributes.description?.raw;
        const typeRaw = literalString(parsed.attributes.type) ?? parsed.attributes.type?.raw;
        const sensitive = literalBoolean(parsed.attributes.sensitive);
        const nullable = literalBoolean(parsed.attributes.nullable);
        const validation = this.extractValidation(parsed.blocks);

        // Parse type constraint if present
        const typeConstraint = typeRaw ? parseTypeConstraint(typeRaw) : undefined;

        return {
            name,
            description,
            type: typeRaw,
            typeConstraint,
            default: parsed.attributes.default,
            validation,
            sensitive,
            nullable,
            raw: block.raw,
            source: block.source
        };
    }

    /**
     * Extracts validation rules from nested validation blocks.
     * @param blocks - Nested blocks from the variable body
     * @returns VariableValidation if a validation block exists
     */
    private extractValidation(blocks: ReturnType<typeof parseBlockBody>['blocks']): VariableValidation | undefined {
        const validationBlock = blocks.find((child) => child.type === 'validation');
        if (!validationBlock) {
            return undefined;
        }

        const condition = validationBlock.attributes.condition;
        const errorMessage = validationBlock.attributes.error_message;

        if (!condition && !errorMessage) {
            return undefined;
        }

        return {
            condition,
            error_message: errorMessage
        };
    }
}

/**
 * Parses a Terraform type constraint expression into a structured TypeConstraint.
 *
 * Supports:
 * - Primitive types: string, number, bool, any
 * - Collection types: list(T), set(T), map(T)
 * - Structural types: object({ attr = type, ... }), tuple([type, ...])
 * - Optional attributes: optional(type)
 *
 * @param raw - The raw type expression string
 * @returns Parsed TypeConstraint
 *
 * @example
 * ```typescript
 * parseTypeConstraint('string')
 * // { base: 'string', raw: 'string' }
 *
 * parseTypeConstraint('list(string)')
 * // { base: 'list', element: { base: 'string', raw: 'string' }, raw: 'list(string)' }
 *
 * parseTypeConstraint('object({ name = string, age = optional(number) })')
 * // { base: 'object', attributes: { name: {...}, age: {...} }, raw: '...' }
 * ```
 */
export function parseTypeConstraint(raw: string): TypeConstraint {
    const trimmed = raw.trim();

    // Primitive types
    const primitives = ['string', 'number', 'bool', 'any'];
    if (primitives.includes(trimmed)) {
        return { base: trimmed, raw: trimmed };
    }

    // Check for collection types: list(T), set(T), map(T)
    const collectionMatch = matchWrappedKeyword(trimmed, ['list', 'set', 'map']);
    if (collectionMatch) {
        const { keyword: base, inner } = collectionMatch;
        return {
            base,
            element: parseTypeConstraint(inner),
            raw: trimmed
        };
    }

    // Check for optional(T)
    const optionalInner = extractWrappedType(trimmed, 'optional');
    if (optionalInner !== null) {
        const inner = parseTypeConstraint(optionalInner);
        return {
            ...inner,
            optional: true,
            raw: trimmed
        };
    }

    // Check for tuple([T1, T2, ...])
    const tupleMatch = trimmed.match(/^tuple\s*\(\s*\[([\s\S]*)\]\s*\)$/);
    if (tupleMatch) {
        const elements = parseTupleElements(tupleMatch[1]);
        const result: TypeConstraint = {
            base: 'tuple',
            raw: trimmed
        };
        if (elements.length > 0) {
            result.elements = elements;
        }
        return result;
    }

    // Check for object({ attr = type, ... })
    const objectMatch = trimmed.match(/^object\s*\(\s*\{([\s\S]*)\}\s*\)$/);
    if (objectMatch) {
        const attributes = parseObjectTypeAttributes(objectMatch[1]);
        return {
            base: 'object',
            attributes,
            raw: trimmed
        };
    }

    // Default: treat as unknown/complex type expression
    return { base: trimmed, raw: trimmed };
}

/**
 * Attempts to extract the inner portion of a keyword-wrapped type expression (e.g., list(...)).
 * @param trimmed - The trimmed raw type string
 * @param keywords - Keywords to try matching against
 * @returns Matched keyword and inner content if found, otherwise null
 */
function matchWrappedKeyword(trimmed: string, keywords: string[]): { keyword: string; inner: string } | null {
    for (const keyword of keywords) {
        const inner = extractWrappedType(trimmed, keyword);
        if (inner !== null) {
            return { keyword, inner };
        }
    }

    return null;
}

/**
 * Extracts inner content from a keyword-wrapped expression while respecting nested parentheses.
 * Returns null if the pattern does not match exactly or parentheses are unbalanced.
 * @param trimmed - The trimmed raw type string
 * @param keyword - The keyword to match (e.g., "list")
 * @returns Inner content if matched, otherwise null
 */
function extractWrappedType(trimmed: string, keyword: string): string | null {
    if (!trimmed.startsWith(keyword)) {
        return null;
    }

    let index = keyword.length;

    while (index < trimmed.length && isWhitespace(trimmed[index])) {
        index++;
    }

    if (index >= trimmed.length || trimmed[index] !== '(') {
        return null;
    }

    index++;
    let depth = 1;
    const start = index;

    while (index < trimmed.length && depth > 0) {
        const char = trimmed[index];
        if (char === '(') {
            depth++;
        } else if (char === ')') {
            depth--;
        }
        index++;
    }

    if (depth !== 0) {
        return null;
    }

    const inner = trimmed.slice(start, index - 1).trim();
    const remainder = trimmed.slice(index).trim();

    if (remainder.length > 0) {
        return null;
    }

    return inner;
}

/**
 * Parses tuple element types from the inner content of a tuple type.
 * @param inner - The content inside tuple([ ... ])
 * @returns Array of TypeConstraints for each element
 */
function parseTupleElements(inner: string): TypeConstraint[] {
    const elements: TypeConstraint[] = [];
    const trimmed = inner.trim();

    if (!trimmed) {
        return elements;
    }

    // Split by commas, respecting nested structures
    let depth = 0;
    let current = '';
    const entries: string[] = [];

    for (let i = 0; i < trimmed.length; i++) {
        const char = trimmed[i];

        if (char === '(' || char === '{' || char === '[') {
            depth++;
            current += char;
        } else if (char === ')' || char === '}' || char === ']') {
            depth--;
            current += char;
        } else if (char === ',' && depth === 0) {
            const entry = current.trim();
            if (entry) {
                entries.push(entry);
            }
            current = '';
        } else {
            current += char;
        }
    }

    // Don't forget the last entry
    const final = current.trim();
    if (final) {
        entries.push(final);
    }

    // Parse each element type
    for (const entry of entries) {
        elements.push(parseTypeConstraint(entry));
    }

    return elements;
}

/**
 * Parses object type attributes from the inner content of an object type.
 * @param inner - The content inside object({ ... })
 * @returns Record of attribute names to their TypeConstraints
 */
function parseObjectTypeAttributes(inner: string): Record<string, TypeConstraint> {
    const attributes: Record<string, TypeConstraint> = {};
    const trimmed = inner.trim();

    if (!trimmed) {
        return attributes;
    }

    // Simple parsing: split by commas (handling nested parentheses)
    let depth = 0;
    let current = '';
    const entries: string[] = [];

    for (let i = 0; i < trimmed.length; i++) {
        const char = trimmed[i];

        if (char === '(' || char === '{' || char === '[') {
            depth++;
            current += char;
        } else if (char === ')' || char === '}' || char === ']') {
            depth--;
            current += char;
        } else if (char === ',' && depth === 0) {
            entries.push(current.trim());
            current = '';
        } else {
            current += char;
        }
    }

    if (current.trim()) {
        entries.push(current.trim());
    }

    // Parse each entry (format: "name = type" or "name = optional(type)")
    for (const entry of entries) {
        const equalsIndex = entry.indexOf('=');
        if (equalsIndex === -1) {
            continue;
        }

        const attrName = entry.slice(0, equalsIndex).trim();
        const attrType = entry.slice(equalsIndex + 1).trim();

        if (!/^\w+$/.test(attrName) || !attrType) {
            continue;
        }

        attributes[attrName] = parseTypeConstraint(attrType);
    }

    return attributes;
}

/**
 * Determines if a character is whitespace without regex backtracking.
 * @param char - The character to evaluate
 * @returns True if whitespace
 */
function isWhitespace(char: string): boolean {
    return char === ' ' || char === '\t' || char === '\n' || char === '\r' || char === '\f';
}

File: python/src/parse_hcl/parsers/variable_parser.py
"""
Parser for Terraform variable blocks.

Extracts variable declarations with type constraints, defaults, and validations.
"""

from __future__ import annotations

import re
from typing import Any, Dict, List, Optional, Tuple

from ..types import HclBlock, TypeConstraint, Value, VariableValidation
from ..utils.common.value_helpers import literal_boolean, literal_string
from ..utils.parser.body_parser import parse_block_body

# Regex patterns for type constraint parsing (matches TypeScript implementation)
COLLECTION_PATTERN = re.compile(r"^(list|set|map)\s*\(\s*([\s\S]*)\s*\)$")
OPTIONAL_PATTERN = re.compile(r"^optional\s*\(\s*([\s\S]*)\s*\)$")
TUPLE_PATTERN = re.compile(r"^tuple\s*\(\s*\[([\s\S]*)\]\s*\)$")
OBJECT_PATTERN = re.compile(r"^object\s*\(\s*\{([\s\S]*)\}\s*\)$")


class VariableParser:
    """
    Parser for Terraform variable definition blocks.

    Example HCL:
        variable "instance_type" {
          type        = string
          default     = "t2.micro"
          description = "EC2 instance type"
          sensitive   = false

          validation {
            condition     = can(regex("^t[23]\\\\.", var.instance_type))
            error_message = "Must be a t2 or t3 instance type."
          }
        }
    """

    def parse(self, block: HclBlock) -> Dict[str, object]:
        """
        Parses a variable block into a structured VariableBlock.

        Args:
            block: The raw HCL block to parse.

        Returns:
            Parsed VariableBlock with all extracted fields including:
            - name: Variable name
            - description: Optional description
            - type: Type constraint expression
            - typeConstraint: Parsed type constraint
            - default: Default value
            - validation: Validation rules
            - sensitive: Whether the variable is sensitive
            - nullable: Whether the variable is nullable
        """
        name = block["labels"][0] if block["labels"] else "unknown"
        parsed = parse_block_body(block["body"])

        description_val = parsed["attributes"].get("description")  # type: ignore[index]
        description = literal_string(description_val) or (description_val.get("raw") if isinstance(description_val, dict) else None)
        type_val = parsed["attributes"].get("type")  # type: ignore[index]
        type_raw = literal_string(type_val) or (type_val.get("raw") if isinstance(type_val, dict) else None)
        sensitive = literal_boolean(parsed["attributes"].get("sensitive"))  # type: ignore[index]
        nullable = literal_boolean(parsed["attributes"].get("nullable"))  # type: ignore[index]
        validation = self._extract_validation(parsed["blocks"])  # type: ignore[arg-type]
        type_constraint = parse_type_constraint(type_raw) if type_raw else None

        return {
            "name": name,
            "description": description,
            "type": type_raw,
            "typeConstraint": type_constraint,
            "default": parsed["attributes"].get("default"),  # type: ignore[index]
            "validation": validation,
            "sensitive": sensitive,
            "nullable": nullable,
            "raw": block["raw"],
            "source": block["source"],
        }

    def _extract_validation(self, blocks: List[Dict[str, object]]) -> Optional[VariableValidation]:
        """
        Extracts validation rules from nested validation blocks.

        Args:
            blocks: Nested blocks from the variable body.

        Returns:
            VariableValidation if a validation block exists, None otherwise.
        """
        validation_block = next((b for b in blocks if b.get("type") == "validation"), None)
        if not validation_block:
            return None

        attrs = validation_block.get("attributes", {})
        condition = attrs.get("condition") if isinstance(attrs, dict) else None
        error_message = attrs.get("error_message") if isinstance(attrs, dict) else None

        if not condition and not error_message:
            return None

        return {"condition": condition, "error_message": error_message}


def parse_type_constraint(raw: str) -> TypeConstraint:
    """
    Parses a Terraform type constraint expression into a structured TypeConstraint.

    Uses regex patterns for robust parsing that handles whitespace correctly.

    Supports:
    - Primitive types: string, number, bool, any
    - Collection types: list(T), set(T), map(T)
    - Structural types: object({ attr = type, ... }), tuple([type, ...])
    - Optional attributes: optional(type)

    Args:
        raw: The raw type expression string.

    Returns:
        Parsed TypeConstraint with base type and nested structure.

    Example:
        >>> parse_type_constraint('string')
        {'base': 'string', 'raw': 'string'}

        >>> parse_type_constraint('list(string)')
        {'base': 'list', 'element': {'base': 'string', ...}, 'raw': 'list(string)'}

        >>> parse_type_constraint('list ( string )')  # Handles whitespace
        {'base': 'list', 'element': {'base': 'string', ...}, 'raw': 'list ( string )'}

        >>> parse_type_constraint('tuple([string, number])')
        {'base': 'tuple', 'elements': [...], 'raw': 'tuple([string, number])'}
    """
    trimmed = raw.strip()
    primitives = {"string", "number", "bool", "any"}

    if trimmed in primitives:
        return {"base": trimmed, "raw": trimmed}

    # Check for collection types: list(T), set(T), map(T) - using regex for whitespace
    collection_match = COLLECTION_PATTERN.match(trimmed)
    if collection_match:
        base, inner = collection_match.groups()
        return {
            "base": base,
            "element": parse_type_constraint(inner.strip()),
            "raw": trimmed,
        }

    # Check for optional(T) - using regex for whitespace
    optional_match = OPTIONAL_PATTERN.match(trimmed)
    if optional_match:
        inner = parse_type_constraint(optional_match.group(1).strip())
        result = dict(inner)
        result["optional"] = True
        result["raw"] = trimmed
        return result  # type: ignore[return-value]

    # Check for tuple([T1, T2, ...]) - preserve element types
    tuple_match = TUPLE_PATTERN.match(trimmed)
    if tuple_match:
        inner_content = tuple_match.group(1).strip()
        elements = _parse_tuple_elements(inner_content)
        result: TypeConstraint = {
            "base": "tuple",
            "raw": trimmed,
        }
        if elements:
            result["elements"] = elements  # type: ignore[typeddict-unknown-key]
        return result

    # Check for object({ attr = type, ... }) - using regex for whitespace
    object_match = OBJECT_PATTERN.match(trimmed)
    if object_match:
        inner_content = object_match.group(1)
        attributes = _parse_object_type_attributes(inner_content)
        result = {
            "base": "object",
            "raw": trimmed,
        }
        if attributes:
            result["attributes"] = attributes
        return result

    # Default: treat as unknown/complex type expression
    return {"base": trimmed, "raw": trimmed}


def _parse_tuple_elements(inner: str) -> List[TypeConstraint]:
    """
    Parses tuple element types from the inner content of a tuple type.

    Handles nested types correctly by tracking bracket depth.

    Args:
        inner: The content inside tuple([ ... ]).

    Returns:
        List of TypeConstraint for each element.
    """
    elements: List[TypeConstraint] = []
    trimmed = inner.strip()

    if not trimmed:
        return elements

    # Split by commas, respecting nested structures
    depth = 0
    current: List[str] = []
    entries: List[str] = []

    for char in trimmed:
        if char in "({[":
            depth += 1
            current.append(char)
        elif char in ")}]":
            depth -= 1
            current.append(char)
        elif char == "," and depth == 0:
            entry = "".join(current).strip()
            if entry:
                entries.append(entry)
            current = []
        else:
            current.append(char)

    # Don't forget the last entry
    final = "".join(current).strip()
    if final:
        entries.append(final)

    # Parse each element type
    for entry in entries:
        elements.append(parse_type_constraint(entry))

    return elements


def _parse_object_type_attributes(inner: str) -> Dict[str, TypeConstraint]:
    """
    Parses object type attributes from the inner content of an object type.

    Handles nested types and complex expressions correctly.

    Args:
        inner: The content inside object({ ... }).

    Returns:
        Record of attribute names to their TypeConstraints.
    """
    attributes: Dict[str, TypeConstraint] = {}
    trimmed = inner.strip()

    if not trimmed:
        return attributes

    # Split by commas, respecting nested structures
    depth = 0
    current: List[str] = []
    entries: List[str] = []

    for char in trimmed:
        if char in "({[":
            depth += 1
            current.append(char)
        elif char in ")}]":
            depth -= 1
            current.append(char)
        elif char == "," and depth == 0:
            entry = "".join(current).strip()
            if entry:
                entries.append(entry)
            current = []
        else:
            current.append(char)

    # Don't forget the last entry
    final = "".join(current).strip()
    if final:
        entries.append(final)

    # Parse each entry (format: "name = type")
    for entry in entries:
        match = re.match(r"^(\w+)\s*=\s*([\s\S]+)$", entry)
        if match:
            attr_name, attr_type = match.groups()
            attributes[attr_name] = parse_type_constraint(attr_type.strip())

    return attributes
